Framestreams

wcawijngaards opened the pull request:
Hi,
This pull request exists to solicit review on the framestreams dnstap branch.  It implements dnstap unidirectional connectivity in unbound.  This has a number of new features.

The dependency on libfstrm is removed.  The fstrm protocol code resides in `dnstap/dnstap_fstrm.h` and `dnstap/dnstap_fstrm.c`.  This contains a brief definition of what unbound needs.

The `make unbound-dnstap-socket` builds a debug tool, unbound-dnstap-socket.  It can listen, accept multiple DNSTAP streams and print information.  Commandline options control it.

Unbound can reconnect if the unix domain socket file socket is closed.  This uses exponential backoff after which it uses a one second timer to throttle cpu down.  There is also support to use TCP and TLS for connecting to the log server.  There are new config options to turn them on, in the dnstap section in the man page and example config file.  `dnstap-ip` with IP address of server for TCP or TLS use. `dnstap-tls` to turn on TLS. And `dnstap-tls-server-name`, `dnstap-tls-cert-bundle`, `dnstap-tls-client-key-file` and `dnstap-tls-client-cert-file` to configure the certificates for server authentication and client authentication, or leave at "" to not use that.

Best regards, Wouter

wcawijngaards committed:
git branch to track feature for framestreams.  The aim is to make
reconnecting work, TLS support and not depend on the libfstrm library,
but keep compatibility with the Frame Streams protocol spec for
existing DNSTAP tools.

wcawijngaards committed:
Writeup of FrameStreams protocol.

wcawijngaards committed:
iolist item added.

wcawijngaards committed:
unbound-dnstap-socket debug test program.

wcawijngaards committed:
dnstap test progam can log on one line type,ip,qname,qtype,qclass

wcawijngaards committed:
dnstap test program prints identity and version.

wcawijngaards committed:
dnstap test program prints messages and timestamps in long format.

wcawijngaards committed:
dnstap unit test.

wcawijngaards committed:
the framestream queue.

wcawijngaards committed:
iothread work.

wcawijngaards committed:
iothread find msg.

wcawijngaards committed:
Remove editor file that was added by mistake.

wcawijngaards committed:
fstrm routines for start and stop, fstrm_create_control_frame_start and
fstrm_create_control_frame_stop, suitable for reuse, together with fstrm
protocol defines.

wcawijngaards committed:
improve fstrm routine documentation.

wcawijngaards committed:
remove fatal_exit from runtime reopen routines.

wcawijngaards committed:
write data and stop flush for dnstap io.

wcawijngaards committed:
output event del before free.

wcawijngaards committed:
dnstap io output performs nonblocking connect.

wcawijngaards committed:
dnstap io fixup socket creation, stop flush dtio pointer, connect reports.

wcawijngaards committed:
dtio_find_msg loop roundrobin instead of first queue only, with
state in the dtio struct for loop iterator.

wcawijngaards committed:
dnstap io: check for close of channel by the other side.

wcawijngaards committed:
check close neater also with a write at the same time.

wcawijngaards committed:
loop in output callback for performance.

wcawijngaards committed:
dnstap io, make sure to free current message when stream closes.

wcawijngaards committed:
dnstap io, sleeps thread when there is no traffic.

wcawijngaards committed:
dnstap io, fixup error exit of event add routines.

wcawijngaards committed:
dnstap io, reconnect attempts with exponential backoff to once per second.

wcawijngaards committed:
in stop flush use timer_del to remove timer.

wcawijngaards committed:
dnstap io, failure to add event closes and attempts to reopen the output.

wcawijngaards committed:
dnstap io, remove libfstrm calls from dnstap.c.

wcawijngaards committed:
dnstap io, remove --with-libfstrm, it is not required to build dnstap
support. protobuf-c is still used for handling the dnstap encoding.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap io, fix event reference in winevent wouldblock call.

wcawijngaards committed:
dnstap io, nicer layout for wouldblock calls.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
merge master into framestreams, autoconf again.

wcawijngaards committed:
dnstap io, fix for msg queue cleanup and make test.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap io, fix memory leak if dnstap queue full.

wcawijngaards committed:
dnstap io, windows portability improvements.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap io, fix exit when compiled without threads.

wcawijngaards committed:
dnstap io, without threads, logs from the main event loop.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap io, fix compile warning when compiled with threading enabled.

wcawijngaards committed:
dnstap io, move setup with nothreads into its own routine.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap socket tool better help text.

wcawijngaards committed:
fix ascync test compile with lock checks.

wcawijngaards committed:
fix dnstap io for lock checks, log identity (numworkers+1), and add locks
around protected area modification in the message queue.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap io, test threads in unit test.

wcawijngaards committed:
dnstap io, add reconnect test.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
- put fstrm protocol contents in separate files, dnstap_fstrm.c
  and dnstap_fstrm.h

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap unbound-dnstap-sock, can listen to multiple sockets, can listen
to TCP sockets, cleans up on exit after signal.

wcawijngaards committed:
dnstap unbound-dnstap-sock, fixup signal handler exit.

wcawijngaards committed:
dnstap unbound-dnstap-sock, fixup constant defines.

wcawijngaards committed:
dnstap unbound-dnstap-sock, tls options and context created.

wcawijngaards committed:
dnstap unbound-dnstap-sock, fixup check for ssl context create error.

wcawijngaards committed:
dnstap unbound-dnstap-sock, add -t option.

wcawijngaards committed:
dnstap unbound-dnstap-sock, verbose accepted stream IP addresses

wcawijngaards committed:
dnstap unbound-dnstap-sock, read from TLS.

wcawijngaards committed:
dnstap unbound-dnstap-sock, comments and log output on tls error close.

wcawijngaards committed:
dnstap io, connect and write over TCP.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap create debug tool with other debug tools in list.

wcawijngaards committed:
dnstap io, ssl and ssl ctx creation.

wcawijngaards committed:
dnstap io, close fd routine.

wcawijngaards committed:
dnstap io, ssl handshake.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap io, ssl write.

wcawijngaards committed:
dnstap io, set tls auth name in outgoing ssl

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap io, check peer verification in dtstream dtio_ssl_handshake.

wcawijngaards committed:
dnstap io, move peer check into routine.

wcawijngaards committed:
dnstap io, check peer verification in unbound-dnstap-socket tap_handshake.

wcawijngaards committed:
dnstap io, example.conf example, config_file entries for tcp and tls.

wcawijngaards committed:
dnstap io, config entries parse and lex.

wcawijngaards committed:
dnstap io, dnstap tls default is yes, and man page documentation.

wcawijngaards committed:
dnstap io, fix clang analysis warning

wcawijngaards committed:
dnstap io, fix to compile without ssl.

wcawijngaards committed:
Nicer comment text.

wcawijngaards committed:
dnstap io, test for TCP and reconnect for that.

wcawijngaards committed:
dnstap io, test for TLS and reconnect for that.  And fix unused parameters
for dt_create and fix check of socket path when using IP address.

wcawijngaards committed:
dnstap io, fix spinning reconnect when handshake fails for TLS.

wcawijngaards committed:
dnstap io, test TLS with peername and TLS authentication.

wcawijngaards committed:
Merge branch 'master' into framestreams

wcawijngaards committed:
Fix ifdef of X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS, and
Merge branch 'master' into framestreams

wcawijngaards committed:
dnstap io, test that failed name or auth certificate fails to connect tls.

wcawijngaards committed:
dnstap io, test for client authentication, unbound can send client
authentication credentials, when configured, and unbound-dnstap-socket can
verify the client credentials, and refuses the connection if missing.

wcawijngaards committed:
dnstap io, move control frame ready, accept and log to dnstap_fstrm code.

wcawijngaards committed:
Fix issue reported by clang analyzer.

wcawijngaards committed:
Fix memory leak in error case.  From review.

ralphdolmans commented on daemon/worker.c:
```diff
@@ -1961,6 +1976,16 @@ worker_delete(struct worker* worker)
+		dt_io_thread_stop(worker->dtenv.dtio);
+	}
+	dt_deinit(&worker->dtenv);
+#endif /* USE_DNSTAP */
```
dtenv gets memcpy'd from daemon to all workers (in worker_create), but not free'd per worker

wcawijngaards replied:
Yeah that copies over some shared globals to the env structure for all workers.  Then there is a per-worker initialisation in dt_init.  The memcpy is from the previous dnstap implementation.  I don't see what to fix (valgrind is leak clean).

ralphdolmans commented on dnstap/dtstream.h:
```diff
@@ -0,0 +1,284 @@
+/*
+ * dnstap/dtstream.h - Frame Streams implementation for unbound DNSTAP
```
Not sure this description really covers the file content. It is about dnstap streams (as the filename suggests), the frame streams specific work is in dnstap_fstrm.

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/unbound/pull/164/commits/58b7e732b839f46b20d940d1ed062b7e7bddc8a0 .

gthess commented on util/netevent.h:
```diff
@@ -786,4 +786,9 @@ void comm_point_tcp_win_bio_cb(struct comm_point* c, void* ssl);
 
+#ifdef HAVE_SSL
+/** true if the ssl handshake error has to be squelched from the logs */
+int squelch_err_ssl_handshake(unsigned long err);
```
Don't we need documentation for the above two functions?

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/unbound/commit/f03245c362ffc78d522f4d8143448575a3e80bca .  I just made them public, but adding documentation is a good idea.

gthess commented on util/net_help.c:
```diff
@@ -1191,6 +1217,44 @@ void* outgoing_ssl_fd(void* sslctx, int fd)
+		SSL_set_verify(ssl, SSL_VERIFY_PEER, NULL);
+	}
+#else
+	verbose(VERB_ALGO, "the query has an auth_name, but libssl has no call to perform TLS authentication");
```
I believe it is better to fail here and suggest that if name verification is not needed after all, dnstap-tls-server-name can be explicitly set to "". But I see that this is also used in the outside_network.c and not sure if we want the same behavior there (because the code runs per query). Maybe we can have on-start code that will test connectivity to configured forwarders/stubs?

wcawijngaards replied:
Yes Ralph also commented on this.  It is a previous function and I do not want to fail existing setups.  It depends on having an older SSL library.

wcawijngaards replied:
Fix in https://github.com/NLnetLabs/unbound/pull/164/commits/398e2601454ec26e91a2d5119739545b4eb66054 as discussed.

gthess commented on util/net_help.c:
```diff
@@ -1191,6 +1217,44 @@ void* outgoing_ssl_fd(void* sslctx, int fd)
+#elif defined(HAVE_X509_VERIFY_PARAM_SET1_HOST)
+	/* openssl 1.0.2 has this function that can be used for
+	 * set1_host like verification */
+	if(auth_name) {
```
auth_name is already checked at the start of the function.

wcawijngaards replied:
Yes, I do the if to get a new context in which I can then register a new X509_VERIFY_PARAM* variable. Could be if(1) I guess to, but I liked this more, or write the if at the start later.

gthess replied:
Ah right, I missed that.
Could we use just `{ ... }`  or will it give warnings with other compilers/systems?

wcawijngaards replied:
I do not have an answer for that. So I'll just close the issue. Code could be beautified in other places too if you find something nicer for it.

ralphdolmans commented on dnstap/dtstream.c:
```diff
@@ -0,0 +1,1839 @@
+	}
+	memset(&tv, 0, sizeof(tv));
+	tv.tv_sec = 2;
+	if(ub_timer_add(timer, info.base, &dtio_stop_timer_cb, &info,
```
dtio_stop_timer_cb should be added to fptr_whitelist_event()

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/unbound/pull/164/commits/6a51e9e037549339fadfac5a68526c51f8fc88c0 . Thanks that is obviously necessary.

ralphdolmans commented on dnstap/dtstream.c:
```diff
@@ -0,0 +1,1839 @@
+		return;
+	}
+	stopev = ub_event_new(info.base, dtio->fd, UB_EV_READ |
+		UB_EV_WRITE | UB_EV_PERSIST, &dtio_stop_ev_cb, &info);
```
add callback to fptr_whitelist_event

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/unbound/pull/164/commits/6a51e9e037549339fadfac5a68526c51f8fc88c0 . Thanks for the find.

ralphdolmans commented on dnstap/dtstream.c:
```diff
@@ -0,0 +1,1839 @@
+{
+	struct ub_event* cmdev;
+	fd_set_nonblock(dtio->commandpipe[0]);
+	cmdev = ub_event_new(dtio->event_base, dtio->commandpipe[0],
```
add callback to fptr_whitelist_event

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/unbound/pull/164/commits/6a51e9e037549339fadfac5a68526c51f8fc88c0 . Thanks for the find.

ralphdolmans commented on dnstap/dtstream.c:
```diff
@@ -0,0 +1,1839 @@
+static void dtio_setup_reconnect(struct dt_io_thread* dtio)
+{
+	dtio_reconnect_clear(dtio);
+	dtio->reconnect_timer = ub_event_new(dtio->event_base, -1,
```
add callback to fptr_whitelist_event

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/unbound/pull/164/commits/6a51e9e037549339fadfac5a68526c51f8fc88c0 . Thanks for the find.

ralphdolmans commented on dnstap/dtstream.c:
```diff
@@ -0,0 +1,1839 @@
+	dtio->check_nb_connect = 1;
+
+	/* the EV_READ is to catch channel close, write to write packets */
+	ev = ub_event_new(dtio->event_base, dtio->fd,
```
add callback to fptr_whitelist_event

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/unbound/pull/164/commits/6a51e9e037549339fadfac5a68526c51f8fc88c0 . Thanks for the find.

ralphdolmans commented on util/net_help.c:
```diff
@@ -1191,6 +1217,44 @@ void* outgoing_ssl_fd(void* sslctx, int fd)
+		SSL_set_verify(ssl, SSL_VERIFY_PEER, NULL);
+	}
+#else
+	verbose(VERB_ALGO, "the query has an auth_name, but libssl has no call to perform TLS authentication");
```
Maybe return 0 here? To make sure the user is aware that no authentication will happen (and should be removed from config).

wcawijngaards replied:
Are you sure? This would break people's configuration for upstream to public resolvers if their openssl happens to be older?  I mean this routine is used there mostly, not so much for DNSTAP.

ralphdolmans replied:
Right. I didn't consider the non-dnstap use. That will indeed break current deployments, making it a bad idea.

wcawijngaards replied:
The current code does not change unbound's behaviour for dns over tls. But shares some code with dnstap. I guess we can close it here, unless you want to change the dnstap behaviour (only)?

wcawijngaards replied:
Fix in https://github.com/NLnetLabs/unbound/pull/164/commits/398e2601454ec26e91a2d5119739545b4eb66054 .

gthess commented on dnstap/unbound-dnstap-socket.c:
```diff
@@ -0,0 +1,1350 @@
+		/* remove trailing newline, tabs to spaces */
+		if(buf[0] != 0) buf[strlen(buf)-1]=0;
+		if(strrchr(buf, '\t')) *strrchr(buf, '\t')=' ';
+		if(strrchr(buf, '\t')) *strrchr(buf, '\t')=' ';
```
Based on the comment above one of those need to be '\n'.

wcawijngaards replied:
The newline is removed above, this is the removal of two tabs.  One before type, and a second one before class. Fix in https://github.com/NLnetLabs/unbound/pull/164/commits/85c4e5883189f96e549113ae579d0f671b6799a5 .

ralphdolmans commented on doc/unbound.conf.5.in:
```diff
@@ -2114,6 +2114,78 @@ If this timeout expires Unbound closes the connection, treats it as
 re-establish a new connection later.
 This option defaults to 100 milliseconds.
+.SS DNSTAP Logging Options
+DNSTAP support, when compiled in, is enabled in the \fBdnstap:\fR section.
```
Maybe add that dnstap will create a thread, as this might change the number you want to configure for num-threads.

(I realize this was also the case in the old implementation)

wcawijngaards replied:
Yes good to note. The old implementation did not have documentation, so I added it. Fixed in https://github.com/NLnetLabs/unbound/pull/164/commits/e24d7c64a8e13160d3fee2c6fe9f9cbf952fcc3b .

ralphdolmans reviewed (approved):
Nice code! I added some remarks.

wcawijngaards committed:
Fixup dtstream.h file description comment. And for dtstream.c too.

wcawijngaards committed:
Add dnstap io callbacks to fptr whitelist event.

wcawijngaards committed:
Dnstap io, note that it creates a thread when possible.

ralphdolmans commented on dnstap/dtstream.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * dnstap/dtstream.c - Frame Streams implementation for unbound DNSTAP
+ * dnstap/dtstream.h - Frame Streams thread for unbound DNSTAP
```
dtstream.h -> dtstream.c

wcawijngaards replied:
Yes, you are right, https://github.com/NLnetLabs/unbound/pull/164/commits/5aa8ae510ed0b69cdb6daf8aca1fe00c1d625d82 fixes it.

wcawijngaards committed:
Fix spelling of dtstream.c

gthess commented on dnstap/unbound-dnstap-socket.c:
```diff
@@ -0,0 +1,1350 @@
+	}
+	/* check peer verification */
+	data->ssl_handshake_done = 1;
+	if(!tap_check_peer(data)) {
```
Maybe this is not needed if the SSL_CTX is setup with
`SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL)`.
That code is part of `util/net_help.c`.

I also see that we use the same mode (SSL_VERIFY_PEER) and logic (checking the client certificate's presence after the handshake) for listening TLS connection for unbound eg. unbound-control.
Could we use the `SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT` mode for all the server-side tls connections?

wcawijngaards replied:
No because dns over tls would not work?  I mean, normal DNS clients do not have client certificates, right?

gthess replied:
I wasn't clear on my text, I meant the server side connections where unbound is configured for client certificates. Currently this is only the debug dnstap file above and unbound-control right? In that case the connection could be closed during the handshake and the client on the other side handle that better than just a connection drop.

wcawijngaards replied:
Yes and only those pass verifypem != NULL, so I can set this in net_help.c, listen_sslctx_create.  Fix in https://github.com/NLnetLabs/unbound/pull/164/commits/e13dfc743d6864cff6ae909cbc1026b1b666c888 .

gthess commented on dnstap/unbound-dnstap-socket.c:
```diff
@@ -0,0 +1,1350 @@
+			return 0;
+		}
+	}
+	s->ev = ub_event_new(base, s->fd, UB_EV_READ | UB_EV_PERSIST,
+		s->ev_cb, s);
```
This fails later in the fptr_wlist when using the mini_event.

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/unbound/pull/164/commits/b63032b4dd7660c0e9bd8eac6f9ba8aac4baf94b .  And a lot of work, because it is an event callback in the debug utils.

gthess commented on dnstap/dnstap_fstrm.c:
```diff
@@ -0,0 +1,203 @@
+			break;
+		}
+		if(field_type == FSTRM_CONTROL_FIELD_TYPE_CONTENT_TYPE) {
+			at+=snprintf(buf+at, sizeof(buf)-at, " content-type(");
```
All these `snprintf()`s may potentially write outside of buf given that a control frame may include various content-types. We can either check the return value with the `sizeof(buf)-at` we pass or maybe use `sldns_str_print()`. IIRC it doesn't allow to write outside of the buffer.

wcawijngaards replied:
In commit https://github.com/NLnetLabs/unbound/pull/164/commits/065506d1a8d8200cd96c1fe128d463e07693f04f it uses sldns_str_print.

gthess commented on util/config_file.c:
```diff
@@ -632,6 +633,13 @@ int config_set_option(struct config_file* cfg, const char* opt,
 #ifdef USE_DNSTAP
 	else S_YNO("dnstap-enable:", dnstap)
 	else S_STR("dnstap-socket-path:", dnstap_socket_path)
+	else S_STR("dnstap-ip:", dnstap_ip)
+	else S_YNO("dnstap-tls:", dnstap_tls)
+	else S_STR("dnstap-tls-server-name:", dnstap_tls_server_name)
+	else S_STR("dnstap-tls-cert-bundle:", dnstap_tls_cert_bundle)
+	else S_STR("dnstap-tls-client-key-file:", dnstap_tls_client_key_file)
+	else S_STR("dnstap-tls-client-cert-file:",
+		dnstap_tls_client_cert_file)
```
Do these options (together with the other dnstap options) make sense here? `dt_create()` is called during worker setup.
Now that I think about it I believe the answer may be libunbound (setting the options before first resolution)? But dnstap is only used in daemon mod.

wcawijngaards replied:
Yes I made them to make sure it was there, but you are right, they may not be used. The get options need to be there, for various output prints from unbound-control and unbound-checkconf.  But the set seems not used.

gthess commented on dnstap/dnstap.c:
```diff
@@ -272,19 +242,31 @@ dt_apply_cfg(struct dt_env *env, struct config_file *cfg)
+	if(!dt_io_thread_register_queue(env->dtio, env->msgqueue)) {
+		log_err("malloc failure");
 		return 0;
+	}
```
This eventually will `fatal_exit()` but don't we need a `dt_msg_queue_delete()` also here?

wcawijngaards replied:
Yes that resource should be freed, fix in https://github.com/NLnetLabs/unbound/pull/164/commits/e856908696ba04f5254cbdb83c9745b95bd4fd2c .

gthess commented on dnstap/dtstream.c:
```diff
@@ -0,0 +1,1839 @@
+			cfg->dnstap_socket_path[0]==0) {
+			log_err("dnstap setup failed, because dnstap is "
+				"enabled, but no dnstap-ip and no "
+				"dnstap-socket-path are given");
```
Maybe different error message like `dnstap setup: no dnstap-socket-path for socket connnect` to be uniform with the following error messages.

wcawijngaards replied:
Yes, that is neater.  Let me fix that in https://github.com/NLnetLabs/unbound/pull/164/commits/49622dd51a6de81ab8ed0e19ca639102cc3763c1 .

gthess commented on dnstap/dtstream.c:
```diff
@@ -0,0 +1,1839 @@
+			dtio_reconnect_enable(dtio);
+			return;
+		}
+	}
```
Although `upstream_is_tls` and `upstream_is_unix` are exclusive I would prefer to see this if-code as part of
`else if(dtio->upstream_is_tcp || dtio->upstream_is_tls)` above.

wcawijngaards replied:
Sure, fixed in https://github.com/NLnetLabs/unbound/pull/164/commits/ed5a9ed1b12378f31b099b9722a39beda8099283 .

gthess reviewed (approved):
Nice work! I had collected the remarks in one review item and may collide with what Ralph already sent.

wcawijngaards committed:
Document log check functions.

wcawijngaards committed:
Fixup ssl authentication not available with check for it.

ralphdolmans commented on util/net_help.c:
```diff
@@ -1217,6 +1217,19 @@ void* outgoing_ssl_fd(void* sslctx, int fd)
+#if !defined(HAVE_SSL_SET1_HOST) && !defined(HAVE_X509_VERIFY_PARAM_SET1_HOST)
+	log_err("the query has an auth_name %s, but libssl has no call to "
+		"perform TLS authentication.  Remove that name from config "
+		"or upgrade the ssl crypto library.", auth_name);
```
should return 0 here

wcawijngaards replied:
Yes, https://github.com/NLnetLabs/unbound/pull/164/commits/5b61afd38c7dd7a0368a526ab8dab0bf2fb90039 is the fix.

wcawijngaards committed:
Return 0 when ssl authentication is not available

wcawijngaards committed:
dnstap debug tool, document string change more clearly.

wcawijngaards committed:
dnstap io, fixup fptr_wlist for unbound_dnstap_socket tool.

wcawijngaards committed:
dnstap io, use sldns_str_print to print to string.

wcawijngaards committed:
dnstap io, free alloced resource on fail path.

wcawijngaards committed:
dnstap io, fix uniform error message for no dnstap-socket-path.

wcawijngaards committed:
dnstap io, fix for review comment.

wcawijngaards committed:
dnstap io, cast void unused return value.

wcawijngaards committed:
For incoming ssl context with verifypem != NULL, we can set
SSL_VERIFY_FAIL_IF_NO_PEER_CERT that can reject client
connections without peer cert during the handshake, which is nicer
than just a connection drop to the client (when we then check
for no peer certificate afterwards).

wcawijngaards committed:
Merge branch 'master' into framestreams
Fixed bison and flex conflicts by regenerating the files.
